xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -1,23 +1,145 @@
 from xdsl.dialects.experimental import dmp
-from xdsl.ir import Attribute, OpResult
+from xdsl.ir import Attribute, Block, BlockArgument, Operation, OpResult, Region
+from xdsl.irdl import Operand
 from xdsl.passes import ModulePass
 from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
     PatternRewriter,
     PatternRewriteWalker,
     RewritePattern,
     op_type_rewrite_pattern,
 )
+from xdsl.rewriter import InsertPoint
+from xdsl.transforms.experimental.stencil_tensorize_z_dimension import (
+    BackpropagateStencilShapes,
+)
 from xdsl.utils.hints import isa
 
 
+def get_stencil_accessed_symbols(op: Operand) -> set[Operand]:
+    """
+    Returns the symbols of all stencil accessess by op and all its dependencies.
+    """
+    res: set[Operand] = set()
+    frontier: set[Operand] = set((op,))
+    next: set[Operand] = set()
+    while len(frontier) > 0:
+        for o in frontier:
+            if isinstance(o, OpResult):
+                if isinstance(o.op, csl_stencil.AccessOp):
+                    res.add(o.op.op)
+                else:
+                    next.update(o.op.operands)
+        frontier = next
+        next = set()
+
+    return res
+
+
+@dataclass(frozen=True)
+class RestructureSymmetricReductionPattern(RewritePattern):
+    """
+    Consume data first where that data comes from stencil accesses to `buf`.
+
+    Identifies a pattern of 2 connected binary ops with 3 args, e.g. of the form `(a+b)+c` with different ops and
+    bracketings supported, and attempts to re-structure the order of computation.
+
+    Being in principle similarly to constant folding, the difference is that args are *not* required to be stencil
+    accesses, but could have further compute applied before being passed to the reduction function.
+    Uses helper function `get_stencil_accessed_symbols` to check which bufs are stencil-accessed in each of these args,
+    and to distinguish the following three cases:
+
+     (1) all accesses in an arg tree are to `buf`  - arg should be moved forward in the computation
+     (2) no accesses are to `buf`                  - arg should be moved backward in the computation
+     (3) there's a mix                             - unknown, take any or no action
+
+    If two args are identified that should be moved forward, or two args are identified that should be moved backwards,
+    the computation is restructured accordingly.
+    """
+
+    buf: BlockArgument
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: arith.Addf | arith.Mulf, rewriter: PatternRewriter, /
+    ):
+        # this rewrite requires exactly 1 use which is the same type of operation
+        if len(op.result.uses) != 1 or not isinstance(
+            use := list(op.result.uses)[0].operation, type(op)
+        ):
+            return
+        c_op = use.operands[0] if use.operands[1] == op.result else use.operands[1]
+
+        def rewrite(one: Operand, two: Operand, three: Operand):
```

Comment:
As far as I can tell, `one`, `two` and `three` have to have the same parent for this to work?

Could we assert this to make sure the logic below always works (or breaks loudly)?

Reply:
This should no longer be an issue after the rewrite.